crates/ruff_python_semantic/src/analyze/class.rs

```diff
@@ -121,12 +121,29 @@ pub fn is_enumeration(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -
     })
 }
 
-/// Returns `true` if the given class is a metaclass.
-pub fn is_metaclass(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> bool {
-    any_base_class(class_def, semantic, &|expr| match expr {
+/// Whether or not a class is a metaclass. Constructed by [`is_metaclass`].
+pub enum IsMetaclass {
+    Yes,
+    No,
+    Maybe,
+}
+
+impl From<IsMetaclass> for bool {
+    fn from(value: IsMetaclass) -> Self {
+        matches!(value, IsMetaclass::Yes)
+    }
+}
+
+/// Returns `IsMetaclass::Yes` if the given class is definitely a metaclass,
+/// `IsMetaclass::No` if it's definitely *not* a metaclass, and
+/// `IsMetaclass::Maybe` otherwise.
+pub fn is_metaclass(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> IsMetaclass {
+    let mut maybe = false;
```

Comment:
> I don't really love having to capture maybe in the closure

hmm, yeah, this does feel a bit... icky 😄

I wonder if a better solution (though it's a bit more of a refactor) would be to split up `any_base_class()` into two functions: one that just handles iterating through all superclasses recursively (`iter_base_classes()` or `iter_superclasses()`), and one that checks if a condition holds true for any of the superclasses (that one could still be called `any_base_class()`). That way `is_metaclass()` could use the lower-level `iter_base_classes()` function directly rather than having to awkwardly work around the fact that the quite high-level `any_base_class` function doesn't really have quite the API we'd like it to have here.

WDYT?

Reply:
I think that's a great idea, I'll give it a try!

Reply:
Just some extra data point if it helps to avoid a lot of work: I don't mind that it it captures the variable

Reply:
This was a little harder than I thought. I've added an `iter_base_classes` function, but it's not truly an iterator now since I build up a `Vec` of base classes and then call `into_iter`. I'll keep thinking about this, but at least the API has the shape I was going for.

Reply:
I kind of prefer the old solution. It overall felt simpler and avoiding the extra allocation is nice too

Reply:
Yeah, writing a recursive iterator seems annoyingly hard :(

Sorry for leading you down the wrong track here @ntBre. My bad -- your first solution probably was best!

Reply:
No worries, it was interesting to try! I've reverted to the original.

One other thing I just noticed, should `maybe` also be `true` for the subscript case? I only handled the `type(foo)` example directly from the issue, but should `type[foo]` work as well? That's just a one-line code change, but I can update the tests too.

Reply:
> should `maybe` also be `true` for the subscript case

I don't think so -- `type[str]`, `type[int]`, etc. always behaves the same as a bare `type` in the context of a class's bases, and if a class inherits from `type` then it's always a metaclass

Reply:
Even subscripts that are "invalid" from a typing perspective behave the same way:

```pycon
>>> class bar(type[int, str]): pass
... 
>>> bar.__bases__
(<class 'type'>,)
```